### PR TITLE
Variance and signature type relations

### PR DESCRIPTION
**Pull request by ahejlsberg**
Ports logic for variance computation, signature type relations, and type predicates.

### CODE REVIEW

File: internal/compiler/checker.go [hunk 1]
```diff
@@ -993,6 +1027,9 @@ func (c *Checker) checkExpressionWorker(node *Node, checkMode CheckMode) *Type {
 		return c.checkPropertyAccessExpression(node, checkMode, false /*writeOnly*/)
 	case SyntaxKindBinaryExpression:
 		return c.checkBinaryExpression(node, checkMode)
+	case SyntaxKindTypeAssertionExpression:
+	case SyntaxKindAsExpression:
```
Comment (jakebailey) [inline internal/compiler/checker.go]: This seems like a bug; should be:

```suggestion
	case SyntaxKindTypeAssertionExpression, SyntaxKindAsExpression:
```

Right?

File: internal/compiler/relater.go [hunk 2]
```diff
@@ -949,6 +975,739 @@ func (c *Checker) typeCouldHaveTopLevelSingletonTypes(t *Type) bool {
...
+				}
+				c.outofbandVarianceMarkerHandler = oldHandler
+				if unmeasurable || unreliable {
+					if unmeasurable {
+						variance |= VarianceFlagsUnmeasurable
+					}
+					if unreliable {
+						variance |= VarianceFlagsUnreliable
+					}
+				}
```
Comment (jakebailey) [inline internal/compiler/relater.go]:
```suggestion
				if unmeasurable {
					variance |= VarianceFlagsUnmeasurable
				}
				if unreliable {
					variance |= VarianceFlagsUnreliable
				}
```

(I think this is a straight port though)

File: internal/compiler/checker.go (hunk 1 above)
Comment (jakebailey) [reply internal/compiler/checker.go]: I think there's a lint rule we can enable which will force us to write a comment in a switch/case that explains why the case is empty (which would help this in the future)

**Comment by ahejlsberg** [comment]
I'll fix, give me a couple of minutes.

File: internal/compiler/checker.go (hunk 1 above)
Comment (ahejlsberg) [reply internal/compiler/checker.go]: Yeah, we need a lint rule here or something.

Comment (jakebailey) [reply internal/compiler/checker.go]: I can't find it, so it may just be a custom rule we have to invent.

**Pull request merged**